Lesson 4

MrZ91 opened the pull request:
![2016-03-13 15 30 38](https://cloud.githubusercontent.com/assets/16620724/13728653/a8afadee-e930-11e5-85ac-f5dead7570c1.png)
https://www.pivotaltracker.com/n/projects/1535999

MrZ91 committed:
[#115556271]Users without a subscription are not permitted in lessons
Subscription controller fixed.

MrZ91 committed:
[#115556265] Only users with 'trainer' role can create courses.
Course controller destroy-action fixed.

MrZ91 committed:
[#115556267] Added a list of courses where user participates.

MrZ91 committed:
[#115579333] Added date of lessons.

MrZ91 committed:
[#115579379] Lessons status added.

MrZ91 committed:
[#115579349] Lessons advancement states added.
[#115556271] Fixed 'unless current_user' error.

SergejOvsiannikov commented on app/assets/javascripts/application.js:
```diff
@@ -9,10 +9,10 @@
 //= require jquery
 //= require jquery_ujs
 //= require jquery-ui
+//= require_tree .
```
наличие этой штуки само по себе плохо.

SergejOvsiannikov replied:
а так она идет в самом конце.

MrZ91 replied:
Если `//= require_tree .` расположить до `//= require jquery`, то во всех файлах джаваскрипта,  `$` не будет интерпретироваться как jQuery и скрипт попросту не будет обрабатываться.

SergejOvsiannikov commented on app/controllers/user/courses_controller.rb:
```diff
@@ -8,9 +8,11 @@ class User::CoursesController < User::AuthenticateController
     @course = current_user.courses.new(course_params)
 
-    if @course.save
+    if (current_user.has_role? :trainer) && @course.save
```
тогда уж current_user.has_role?(:trainer) && @course.save

c3gdlk replied:
это должен проверять CanCan

SergejOvsiannikov commented on app/controllers/user/subscriptions_controller.rb:
```diff
@@ -12,14 +12,19 @@ def create
 
   def show
-    @subscriptions = current_user.subscriptions.page(params[:page]).per(COURSES_ON_PAGE)
+    @subscriptions = current_user.subscriptions.page(params[:sbscr]).per(COURSES_ON_PAGE)
```
`params[:sbscr]` переименуй. я понимаю что ты хотел сказать, но это сокращение не читаемо.

SergejOvsiannikov commented on app/models/lesson.rb:
```diff
@@ -20,7 +20,7 @@ def set_position
   def course_lessons_count
-    count = Lesson.where(course_id: course_id).count
+    count = course.lessons.count
     count ? count : 0
```
зачем тут эта строка? count и так вернет 0 если там пусто

SergejOvsiannikov commented on app/views/user/subscriptions/show.html.haml:
```diff
@@ -1,19 +1,47 @@
-  .col-md-10
+%ul#coursetab.nav.nav-tabs
+ %li.active
+  %a{"data-toggle" => "tab", :href => "#subscribtion"} Subscriptions
```
rocket hash :)  и вообще сделай через link_to

SergejOvsiannikov commented on app/views/user/subscriptions/show.html.haml:
```diff
@@ -1,19 +1,47 @@
+ %li.active
+  %a{"data-toggle" => "tab", :href => "#subscribtion"} Subscriptions
+ %li
+  %a{"data-toggle" => "tab", :href => "#participating"} Participating
```
+

SergejOvsiannikov commented on app/views/user/subscriptions/show.html.haml:
```diff
@@ -1,19 +1,47 @@
+       = link_to 'Hidden', '#', class: 'btn btn-danger disabled', style:('display: none' unless !course.visible)
 
+   %br/
+  = paginate @subscriptions, :params => {:anchor => "subscribtion"}, param_name: 'sbscr'
```
+

SergejOvsiannikov commented on app/views/user/subscriptions/show.html.haml:
```diff
@@ -1,19 +1,47 @@
- %br/
-= paginate @subscriptions
+   %br/
+  = paginate participations, :params => {:anchor => "participating"}, param_name: 'prctp'
```
+

SergejOvsiannikov commented on config/initializers/rolify.rb:
```diff
@@ -0,0 +1,2 @@
+Rolify.configure do |config|
```
толку от пустого конфига?

MrZ91 committed:
Fix find_by; Social network auth fix

MrZ91 committed:
Merge branch 'lesson-3' into lesson-4

Conflicts:
	app/controllers/user_controller.rb
	app/views/courses/_preview.html.haml
	lib/modules/omniauthable.rb

MrZ91 committed:
Fixes from lesson-3

MrZ91 committed:
Fixes from lesson-3

MrZ91 committed:
Merge branch 'lesson-4' into lesson-5

Conflicts:
	app/controllers/user/advancements_controller.rb
	app/controllers/user/lessons_controller.rb
	app/models/lesson.rb
	app/views/courses/_lesson.html.haml
	app/views/user/courses/_homework.html.haml
	spec/controllers/user/lessons_controller_spec.rb

MrZ91 committed:
[#115556267] Added a list of courses where user participates.
Restored.

MrZ91 committed:
[#115579429] User newsfeed implemented.
User receiving: advancement status changing; lessons creating; course visibility changing; excluding from course.
User receiving: user subscribing to course; user sending/resending advancement; user.

c3gdlk commented on app/controllers/advancement_controller.rb:
```diff
@@ -1,23 +1,15 @@
-        render 'advancement/error'
-      end
-    end
+    render 'advancement/error' unless current_user.advancements.where(lesson_id: @lesson.id).count.zero?
```
Что делает этот контроллер? Мне название модели ни о чем не говорит.

Эта строчка выглядит очень странно, подумайте как можно сделать еще

c3gdlk commented on app/controllers/application_controller.rb:
```diff
@@ -22,8 +21,13 @@ def select_layout
+    redirect_to signed_up_with_social_user_profile_path
   end
+
+  def profile_configured?
```
этого же не должно быть. Пользователь может регистрироавться без емейла

c3gdlk commented on app/controllers/user/courses_controller.rb:
```diff
@@ -20,6 +22,8 @@ def index
   end
 
   def new
+    redirect_to :back unless current_user.has_role? :trainer
```
CanCan

c3gdlk commented on app/controllers/user/subscriptions_controller.rb:
```diff
@@ -12,14 +12,19 @@ def create
   end
+
+  def participations
+    @participations ||= Kaminari.paginate_array(current_user.participations.distinct).page(params[:particip_page]).per(COURSES_ON_PAGE)
```
похоже на костыль. Зачем Вам дистинкт, должны быть валидации которые не дадут на курс 2 раза подписаться

MrZ91 replied:
Эта реализация до того, как я прочёл Ваш ответ в гугло-группе. Коммит с этой реализацией "потерялся" во время мерджей.
Это коллекция курсов, в которых пользователь "принимает участие" т.е. уже начал выполнять задания к урокам, и она никак напрямую не связана с подпиской. Дистинкт потому, что если заданий в рамках этого курса больше одного current_user.participations возвращает одинаковые копии этого курса.
Я поправлю эту часть, чтобы выводился только список курсов на которые пользователь подписан (реализовал я это ещё в предыдущем ДЗ)

c3gdlk commented on app/helpers/application_helper.rb:
```diff
@@ -9,3 +9,12 @@ def full_title(page_title)
   end
 end
+
+def random_from(n, from)
```
Вы же только в сидах это используете. Зачем метод в хелпере?

c3gdlk commented on app/models/course.rb:
```diff
@@ -13,4 +15,8 @@ class Course < ActiveRecord::Base
   end
+
+  def sort_lessons_by_order(order)
+    lessons.each { |l| l.update(position: (order.index(l.id.to_s) + 1)) }
```
l - это не имя для переменной

c3gdlk commented on app/models/lesson.rb:
```diff
@@ -20,8 +20,7 @@ def set_position
   def course_lessons_count
-    count = Lesson.where(course_id: course_id).count
-    count ? count : 0
+    course.lessons.count
```
зачем уроку знать сколько всего уроков в курсе?

MrZ91 replied:
У каждого урока в курсе есть поле позиция. По умолчанию, при создании урока, значение в этом поле на 1 больше чем кол-во уже созданных курсов. Это поле используется для вывода уроков в заданной последовательности (по умолчанию новые ниже старых), и для выполнения перестановки уроков в этой последовательности.

c3gdlk commented on app/models/user.rb:
```diff
@@ -1,19 +1,26 @@
+
   include Omniauthable
 
+  after_commit :apply_default_role
```
after_commit не запускается в тесте, использовать стоит только в случае необходимости.

c3gdlk commented on app/views/courses/_homework.html.haml:
```diff
@@ -10,5 +10,5 @@
     %th
   %tbody.rowlink{ data: { link: 'row' } }
-   - course.lessons.each do |lesson|
+   - course.lessons.by_position.each do |lesson|
```
почему в паршиале с именем домашнее задание выводятся все уроки?

c3gdlk commented:
где CanCan?

MrZ91 committed:
[#115579413] Newsfeed mailers implement.
Nesfeed background jobs with Sidekiq implement.

MrZ91 committed:
Merge branch 'lesson-3' into lesson-4

MrZ91 committed:
[#115556265] CanCanCan gem add.

MrZ91 committed:
Merge branch 'lesson-4' into lesson-5

MrZ91 committed:
[#115579435] Simple API implement.

c3gdlk commented:
не могу найти изменения aubscriptions  контроллера публичного в котором пользователь подписывается на курс или отписывается от него. Есл пользователь забанен - он не может этого делать

c3gdlk commented:
не могу найти изменения aubscriptions  контроллера публичного в котором пользователь подписывается на курс или отписывается от него. Есл пользователь забанен - он не может этого делать

MrZ91 commented:
Т.е. сделать через авторизацию CanCanCan? В таком случае нужно переопределить обработчик исключения из CanCanCan? У меня просто возникают проблемы с рендером/редиректом после Ajax запросов, можно ли будет сделать обработчик исключений с рендером .js файла?

MrZ91 committed:
CanCanCan authorization for subscriptions contoller.

MrZ91 committed:
[#115579435] User subscription API implement.

c3gdlk commented:
Да, нужно использовать cancancan. Обработчик тоже можно переопределить - использовать respond_with do |format| 

или явно проверять - request.xhr?

c3gdlk commented:
:+1:

MrZ91 committed:
Merge pull request #15 from MrZ91/lesson-5

Lesson 5
